File: bindings/wasm/iota_interaction_ts/lib/move_calls/identity/upgrade.ts
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { SharedObjectRef } from "@iota/iota-sdk/dist/cjs/bcs/types";
import { ObjectRef, Transaction } from "@iota/iota-sdk/transactions";

export function proposeUpgrade(
  identity: SharedObjectRef,
  capability: ObjectRef,
  packageId: string,
  expiration?: number,
): Promise<Uint8Array> {
  const tx = new Transaction();
  const cap = tx.objectRef(capability);
  const identityArg = tx.sharedObjectRef(identity);
  const exp = tx.pure.option('u64', expiration);

  tx.moveCall({
    target: `${packageId}::identity::propose_upgrade`,
    arguments: [identityArg, cap, exp],
  });

  return tx.build();
}

export function executeUpgrade(
  identity: SharedObjectRef,
  capability: ObjectRef,
  proposalId: string,
  packageId: string,
): Promise<Uint8Array> {
  const tx = new Transaction();
  const cap = tx.objectRef(capability);
  const proposal = tx.pure.id(proposalId);
  const identityArg = tx.sharedObjectRef(identity);

  tx.moveCall({
    target: `${packageId}::identity::execute_upgrade`,
    arguments: [identityArg, cap, proposal],
  });

  return tx.build();
}

File: identity_iota_core/src/rebased/iota/move_calls/identity/upgrade.rs
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use iota_interaction::rpc_types::OwnedObjectRef;
use iota_interaction::types::base_types::ObjectID;
use iota_interaction::types::base_types::ObjectRef;
use iota_interaction::types::programmable_transaction_builder::ProgrammableTransactionBuilder;
use iota_interaction::types::transaction::ObjectArg;
use iota_interaction::types::transaction::ProgrammableTransaction;
use iota_interaction::ident_str;

use crate::rebased::iota::move_calls::utils;

pub(crate) fn propose_upgrade(
  identity: OwnedObjectRef,
  capability: ObjectRef,
  expiration: Option<u64>,
  package_id: ObjectID,
) -> Result<ProgrammableTransaction, anyhow::Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let cap_arg = ptb.obj(ObjectArg::ImmOrOwnedObject(capability))?;
  let identity_arg = utils::owned_ref_to_shared_object_arg(identity, &mut ptb, true)?;
  let exp_arg = utils::option_to_move(expiration, &mut ptb, package_id)?;

  let _proposal_id = ptb.programmable_move_call(
    package_id,
    ident_str!("identity").into(),
    ident_str!("propose_upgrade").into(),
    vec![],
    vec![identity_arg, cap_arg, exp_arg],
  );

  Ok(ptb.finish())
}

pub(crate) fn execute_upgrade(
  identity: OwnedObjectRef,
  capability: ObjectRef,
  proposal_id: ObjectID,
  package_id: ObjectID,
) -> Result<ProgrammableTransaction, anyhow::Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let cap_arg = ptb.obj(ObjectArg::ImmOrOwnedObject(capability))?;
  let proposal_id = ptb.pure(proposal_id)?;
  let identity_arg = utils::owned_ref_to_shared_object_arg(identity, &mut ptb, true)?;

  let _ = ptb.programmable_move_call(
    package_id,
    ident_str!("identity").into(),
    ident_str!("execute_upgrade").into(),
    vec![],
    vec![identity_arg, cap_arg, proposal_id],
  );

  Ok(ptb.finish())
}
